src/Compilers/CSharp/Portable/BoundTree/DecisionTree.cs

```diff
@@ -0,0 +1,916 @@
+﻿using Microsoft.CodeAnalysis.CSharp.Symbols;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Roslyn.Utilities;
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    partial class BoundPatternSwitchStatement
+    {
+        private ImmutableArray<Diagnostic> _decisionTreeDiagnostics;
+        private DecisionTree _decisionTree;
+
+        public ImmutableArray<Diagnostic> DecisionTreeDiagnostics
+        {
+            get
+            {
+                EnsureDecisionTree();
+                Debug.Assert(_decisionTree != null);
+                return _decisionTreeDiagnostics;
+            }
+        }
+
+        public DecisionTree DecisionTree
+        {
+            get
+            {
+                EnsureDecisionTree();
+                Debug.Assert(_decisionTree != null);
+                return _decisionTree;
+            }
+        }
+
+        public LabelSymbol ConstantTargetOpt
+        {
+            get
+            {
+                EnsureDecisionTree();
+
+                // Switch expression might be a constant expression.
+                // For this scenario we can determine the target label of the switch statement
+                // at compile time.
+                if (!SwitchSections.Any())
+                {
+                    // empty switch block, set the break label as target
+                    return this.BreakLabel;
+                }
+
+                return BindConstantTarget(_decisionTree);
+            }
+        }
+
+        /// <summary>
+        /// Used to compute the only target label when it can be computed at compile-time.
+        /// This does not use the same logic as handling "goto case n" because that
+        /// always goes to "case n" even if there is a preceding "case n when something".
+        /// However, the initial dispatch always obeys the switch statement sequentual
+        /// semantics. We simulate that here.
+        /// </summary>
+        private LabelSymbol BindConstantTarget(DecisionTree decisionTree)
```

Comment:
I did start out that way, and the code was much messier. I like the fact that this is `private`. I'd rather keep `DecisionTree` closer to the functional style we use for the bound nodes in which it appears.

---
In reply to: [64669055](https://github.com/dotnet/roslyn/pull/11485#discussion_r64669055) [](ancestors = 64669055)